### ISSUE TITLE
Elementary Mathematical Functions of Purely Imaginary Arguments

### ISSUE THREAD

**Issue by damianmoz** [chapel-lang/chapel#24506]
These are vast simplifications of the **complex**(w) ones that should be crafted at some stage to ensure completeness of the mathematical library.  Relying on an implicit conversion of **imag** to a **complex** to be able to use the **complex** implementation of these functions condemns the user to doing significantly more computations. Many of them return a **complex** number,  some return an **imag** number, and others return a **real** number. Note that their existence in C is noted in Appendix G of the C standard. However, no C `libm` library (of which I know) implements them - yet. Note that this is bound to change when the new standard comes out during 2024??? Chapel's implementation of the arithmetic of purely imaginary numbers is more complete that that of any C compiler as of January 2024.

Chapel has not defined these (except for **abs** and **conj**). Routines like the complex square root of a purely imaginary number, the phase, or the projection onto the extended complex plane, are either trivial or quite simple.

Many are what could be called reasonably simple. For example, given an **imag**(w) number x, and
```chapel
z = exp(x);
```
then `z.re == cos(x)` and `z.im == sin(x)`. The two components should be computed with the same low overhead as that of the non-standard `sincos` routine seen in the **gcc** or **llvm** C libraries.

The following definitions hold:
```chapel
cos(i y) = cosh(y)
sin(i y) = i sinh(y)
tan(i y) = i tanh(y)
cosh(i y) = cos(y)
sinh(i y) = i sin(y)
tanh(i y) = i tan(y)
asin(i y) = i asinh(y)
atan(i y) = i atanh(y)
asinh(i y) = i asin(y)
atanh(i y) = i atan(y)
```
A more detailed proposal is needed (it will be attached - but not now). There are obviously edge cases which need to be addressed in that document.

Priority - Low. Related to #24490, #24464, and probably others. Apologies for any omissions.

**Comment by bradcray** [comment]
@damianmoz / @lydia-duncan : Correct me if I'm wrong, but as I understand it, any of these routines that would return a `real` or `imag` rather than a `complex` would need to be defined (or at least made into errors / marked as unstable) before 2.0 in order for us to change them to that return type in the future, right?  Oh, I guess that's assuming they return a `complex` today...  do they?

If that's the case, and this is low-hanging fruit, then it'd be nice to do something here prior to the 2.0 release.

**Comment by bradcray** [comment]
(noting for myself that if we had a list of all such procedures, a very easy solution would be to just define them like:

```chapel
@unstable(...)
proc foo(x: imag) {
  return foo(x: complex);
}
```

to preserve the current implementation while marking them as unstable.  And that we could find a superset of them by looking for procedures that accept and return complex today.  But we'd only need to mark the ones that had a narrower return type, which is harder to look for in an automated fashion).

**Comment by damianmoz** [comment]
Note sure I understand or more importantly, whether I have explained things clearly

As far as I can tell, nowhere are there defined today routines like
```chapel
proc foo(x : imag(?w))
```
that need to be marked as unstable as far as I can tell.

The only routines that are defined for **imag**(w) types today are `conj()` and `abs()` and they are defined both correctly and completely. Two **fma**() routines are defined on **imag**(w) types but they are already marked as _unstable_.  Note I am referring to the nightly snapshot from about two weeks ago.

I was mentioning these other routines now as @mppf suggested to put a marker on the wall into which somebody can stick a pin at some stage in the future. If that work starts **before** 2.0, then I agree with you that they should be marked as _unstable_. But as nobody is doing it yet, it is a non-issue (so to speak). I was not trying to put any more work on anyone's plate before 2.0

**Comment by damianmoz** [comment]
I am curious why both **conj**() and **abs**() passed an **imag**(w) argument need a **pragma** to handle the case of a GPU and CPU

**Comment by lydia-duncan** [comment]
Brad's point is that because it's possible to make calls today using `imag` and get a `complex` back, changing the return type later may break people's code if they are relying on the type returned being `complex` and we change it in the future.  So making an explicit overload will allow us to mark that case as unstable so that at least people that are relying on the return type will get a warning if they compile with `--warn-unstable`

**Comment by jabraham17** [comment]
I am making a potential list of functions that would need to be marked unstable prior to 2.0. I looked at all functions in `Math` and `AutoMath` that accept a `complex` argument and have no `imag` argument overload.

These take a `complex` argument and return a complex argument. They need an overload for `imag` that is marked unstable since the return type may change
```
acos
acosh
asin
asinh
atan
atanh
cos
cosh
sin
sinh
tan
tanh
```

The following take a `complex` argument with no `imag` overload, but I don't think they do not need an explicit unstable `imag` overload. For example, `exp(i)` is a complex number with both real and imaginary parts
```
exp
ln
log
sqrt
```

@damianmoz if this list looks accurate to you, I add unstable imaginary overloads of the trig functions listed (essentially just the ones listed [here](https://chapel-lang.org/docs/main/modules/standard/Math.html#trigonometric-functions)).

**Comment by jabraham17** [comment]
Ok, I will exclude `acos` and `acosh` from the list.

> Also, in the documentation, you incorrectly group the hyperbolic functions amongst the trigonometric functions

That sounds like a simple documentation fix to just add a `Hyperbolic Functions` section to the Math docs.

**Comment by e-kayrakli** [comment]
> I am curious why both conj() and abs() passed an imag(w) argument need a pragma to handle the case of a GPU and CPU

@damianmoz - I am not sure I follow. `conj` and `abs` that take `imag` arguments don't seem to have pragmas.

You're probably referring to `pragma "codegen for CPU and GPU"`. While serving the same purpose, you could see that pragma in two different cases:

1. On a `proc`. If this is the case, it might be a relic from earlier days of GPU support. Currently, the compiler can mark Chapel procs for GPU codegen automatically if necessary.
2. On an `extern proc`. These are still necessary. Though, we can replace them with an attribute or make things work without any sort of annotation. Design-wise, this is a bit trickier then (1). Because, currently this pragma tells the compiler that there will be a `__device__` function that we can link with during linkage. Removing any annotation would probably require a good error message for when such a function doesn't exist and clearer documentation. As a compiler implementer, I like having some sort of annotation as a "contract" between the compiler and the user, which can also serve as a reminder to the user to make sure they know what they are doing. As a user, I _might_ find it cumbersome.

Note that `pragma`s are never user-facing. We'll definitely have to do something about that.

**Comment by damianmoz** [comment]
My brain was trying to run in symmetric multiprocessing mode with insufficient cores last night when I posted that. Sorry Engin.

I was actually looking at the `sqrt` routine in my previous post.

Given a **param** argument, this calls the `sqrt` primitive which does things at compile time.

Given a **const** argument, the **proc** `sqrt` (which does have those **pragma**s) just calls the libm C routines. Using `--fast` with the Chapel compiler, the `-no-math-errno` flag is also fed to the underlying C compiler to force the bowels of the compiler to use an assembler instruction or an intrinsic. I think. @mppf - please correct me if I am wrong.

My questions was why, in this latter case, one cannot just use the primitive within the routine. That results in less code, is simpler, and exploits the machine instruction irrespective of the optimization level being used or avoids the use of an extra compiler flag. I think all GPUs have a machine level SQRT instruction.

The reason I ask is that
a) @mppf kindly did this for `sqrt` at my behest in 2023. Thank you Michael.
b) other compiler mathematical intrinsics for other operations is on the cards - #24464.

I wanted to understand all the intricacies of what was done for `sqrt` and what I do not understand before I say something stupid in case (b). Other operations, one of which is say minimum/maximum, often have their own subtleties and may be slightly more complicated compared to `sqrt` (which was a no-brainer).

Thx. I hope my explanation is better at this attempt.

**Comment by damianmoz** [comment]
I have now read your notes about when you need that **pragma**, five times in fact, and I am not any wiser. I might attack it again tomorrow.

**Comment by mppf** [comment]
@damianmoz - I think maybe you aren't looking at `main` ? E.g. `git grep 'extern proc sqrt'` gives no results. Recent PRs have changed this to use the LLVM intrinsic when compiling with the LLVM backend.

**Comment by damianmoz** [comment]
I grabbed a copy of `main` on February 17th thinking that I had all the major changes that affect me and those `extern proc sqrt` references were still in that snapshot. But they are gone from today's `git clone`. You have been busy!!!

Silly (out-of-date) me!

**Comment by lydia-duncan** [comment]
> Also, in the documentation, you incorrectly group the hyperbolic functions amongst the trigonometric functions. But that is another issue.

That'll be fixed by #24569, thanks for pointing it out!

**Comment by jabraham17** [comment]
In case people missed it, noting here that I merged https://github.com/chapel-lang/chapel/pull/24525 to mark the functions this change would affect as unstable

**Comment by damianmoz** [comment]
At some stage, I will post versions **sgn**(y: imag(?w)) and **sgn**(z:complex(?w)) which return a **real**(w) and **real**(w/2) respectively. They are mathematically correct except that they preserve the sign of -0.0 as perr IEEE 754.

Whether they need @unstable is questionable.

**Comment by jabraham17** [comment]
After #24525 marked these functions unstable, we are getting ready to re-stabilize them with the following definitions.

```chapel
 proc cos(x: imag(?w)): real(w) do return cosh(x:real(w));
 proc cosh(x: imag(?w)): real(w) do return cos(x:real(w));

 proc sin(x: imag(?w)): imag(w) do return sinh(x:real(w)) : imag(w);
 proc sinh(x: imag(?w)): imag(w) do return sin(x:real(w)) : imag(w);

 proc tan(x: imag(?w)): imag(w) do return tanh(x:real(w)) : imag(w);
 proc tanh(x: imag(?w)): imag(w) do return tan(x:real(w)) : imag(w);


 proc asin(x: imag(?w)): imag(w) do return asinh(x:real(w)) : imag(w);
 proc asinh(x: imag(?w)): imag(w) do return asin(x:real(w)) : imag(w);

 proc atan(x: imag(?w)): imag(w) do return atanh(x:real(w)) : imag(w);
 proc atanh(x: imag(?w)): imag(w) do return atan(x:real(w)) : imag(w);
```

These are written using in a shorthand with `?w`, but the version that gets implemented will likely fully spell out the types as that matches the rest of the functions in `Math.chpl`. For example:

```chapel
 proc cos(x: imag(32)): real(32) do return cosh(x:real(32));
 proc cos(x: imag(64)): real(64) do return cosh(x:real(64));
```

@damianmoz, do these match your expectations?

**Comment by damianmoz** [comment]
I will get back to you later today my time - Saturday.

**Comment by damianmoz** [comment]
I was quoting those functions directly from the appendix of the C standard. I am late with some promised documentation but I figured I was covered as long as they were unstable. I did not realize that you would be so conscientious.

Why not **inline** the above? But there is more ....

Hijacking some of the incomplete and unchecked content on my ChapelCon24 submission, there is a major mathematical issue with the
```chapel
atan(y : imag(?w)); // returns a number of the form 0+iY only for |y| <= 1
asinh(y : imag(?w)); // returns a number of the form 0+iY only for |y| <= 1
```
Except for a very limited range of _y_ (as noted in the comments), the mathematical function has a result with a non-zero real component, i.e. a genuine complex number. This is also an issue in C which is under discussion at the moment. So, these two **proc**s might be more useful if they say
```chapel
inline proc asinh(x: imag(?w)): complex(2 * w) do return asinh(y:complex(2 * w));
inline proc atan(x: imag(?w)): complex(2 * w) do return atan(y:complex(2 * w));
```
There is still an argument for performance reasons to demand that these two return an **imag**(w) result for all values of _y_. This means returning a _NaN_ outside the range for which the mathematical result is purely imaginary. Not my preference when one can easily return the correct numerical result rather than something which has (what I perceive as) limited usefulness.

Maybe the @unstable needs to stay for those two.

You also say
```
These are written using a shorthand with ?w, but the version
that gets implemented will likely fully spell out the types as
that matches the rest of the functions in Math.chpl.
```
I do not understand why. To me, the advantage of Chapel is that one can write
```
inline proc cos(x: imag(?w)): real(w) do return cosh(x:real(32));
```
That halves the number of lines. Once Chapel supports **real**(16) and **real**(128) types, the number of lines saved would be even more. I want to promote generic floating point type thinking.

Actually I would omit the return type from the code but I assume you want that there for the Math.chpl documentation which gets automagically generated from the raw source code.

Also, what of the following two ways of expressing the argument is better
```chapel
0:real(w) + y
y : complex(2 * w)
```
If you understand all of the above, say a few words about the mathematics of purely imaginary numbers and how straightforward it was to write this interface in Chapel, and add some semi-intelligent arm but very very intelligent waving, you can now present my ChapelCon24 talk instead of me :)

**Comment by mppf** [comment]
> > These are written using a shorthand with ?w, but the version that gets implemented will likely fully spell out the types as that matches the rest of the functions in Math.chpl.

> I do not understand why. 

It has to do with implicit conversions. Something with type `complex(?w)` won't allow implicit conversions from the other numeric types (`int(8)` or `imag(32)` or `real(64)` as some example types we could convert to complex). There is no problem at all using `complex(?w)` if you do not wish to allow implicit conversions for the call. (Such implicit conversions are unstable -- they will be allowed by the compiler today but will stop being allowed in the future). See also issue #20011 and PR #22769.

It is moot for `imag(?w)` since the only implicit conversions to an `imag(?w)` type is `imag(32)` -> `imag(64)` and this won't be used (since a `imag(32)` formal would be used if the call is using `imag(32)`).

**Comment by damianmoz** [comment]
Jade, I prefer the correctness approach too.

Michael, Thanks for the insight.

**Comment by jabraham17** [comment]
Noting here that I have opened https://github.com/chapel-lang/chapel/pull/24938 which will implement many of the changes discussed here. `asinh` and `atan` are not included in this PR

**Comment by jabraham17** [comment]
I have also opened https://github.com/chapel-lang/chapel/pull/24999 which will remove the `@unstable` overloads of `asinh` and `atan`. I think merging this will address all of the issues outlined here and we can close this issue.

**Comment by bradcray** [comment]
Thanks for taking care of this, Jade!
